[POS] Simultaneous cash and card payment handling

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13278 13278
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds improvements to the cash payment collection flow:
1. When "Mark payment as complete" is tapped in the Cash payment screen, the card reader payment intent is canceled.
2. If the card payment is initiated, the Cash payment screen should be interrupted/closed automatically, showing the card payment processing state.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
Verify the following on `trunk`:
##### Payment intent cancelation
1. Open POS (menu more > Point of Sale)
2. Create order and make sure the reader is connected
3. Click "Cash payment" 
4. Complete order by clicking the button, and observe that card reader payment action is not canceled

##### Cash payment screen not closed when card payment initiated:
1. Open POS (menu more > Point of Sale)
2. Create order and make sure the reader is connected
3. Click "Cash payment" 
4. Tap card to the reader and observe Cash payment screen is not closed automatically

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

It should be verified that the above scenarios are fixed.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
* Samsung S9 Tab, Android 14
* Verified that payment intent is canceled properly when order is marked completed (with cash)
* Verified that Cash payment screen gets closed automatically when user taps card to the reader

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
1. Payment intent gets canceled when order marked as paid with cash

https://github.com/user-attachments/assets/f5519b00-ef3b-4828-8e01-fe08f40b228f

3. Cash payment is closed when card is tapped to the reader

https://github.com/user-attachments/assets/8370cb93-6a25-49ad-893a-92cb0d6f3ae3


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Cancel payment intent when "Mark ... as complete" tapped

samiuelson committed:
Return to Home from Cash if card payment initiated

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Merge branch 'trunk' into pos-cash-card-payment-collection-alignment

samiuelson committed:
Merge branch 'trunk' into pos-cash-card-payment-collection-alignment

samiuelson committed:
Rename event

samiuelson committed:
Rename event

samiuelson committed:
Reorg imports

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/cashpayment/WooPosCashPaymentNavigation.kt:
```diff
@@ -49,6 +53,18 @@ fun NavGraphBuilder.cashPaymentScreen(
         },
     ) { backStackEntry ->
+        val homeViewModel = hiltViewModel<WooPosHomeViewModel>()
+        LaunchedEffect(homeViewModel) {
```
Do you know if we need to have this code at all? Don't we handle the event in `WooPosHomeScreen` already?

samiuelson replied:
The LaunchedEffect coroutine in WooPosHomeScreen is not active when we navigate to CashPayment screen. That's why I subscribed to the events in the CashPayment screen here again. 

An alternative approach could be to move the subscription to navigation events up in the compose navigation hierarchy, to the`WooPosRootHost`, in order to avoid duplicate subscriptions – 5f9d68ecf2044653d4d32df8213756f8d1d1ce9b. Let me know what you think, @kidinov !

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeViewModel.kt:
```diff
@@ -138,6 +138,7 @@ class WooPosHomeViewModel @Inject constructor(
                         _state.value = _state.value.copy(
                             screenPositionState = ScreenPositionState.Checkout.FullScreenTotals
                         )
+                        _navigationEvent.emit(NavigationEvent.ReturnHomeFromCashPayment)
```
Those `NavigationEvent` are `ChildToParentEvent` events, so semantically, I think it'll be better to send this from `WooPosTotalsViewModel::listenToPaymentState`

samiuelson replied:
Done – 6259911cd85de2bb6ceb80b84958c50c3e058656

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeChildToParentCommunication.kt:
```diff
@@ -43,6 +43,7 @@ sealed class ChildToParentEvent {
     sealed class NavigationEvent : ChildToParentEvent() {
         data class ToCashPayment(val orderId: Long) : NavigationEvent()
         data class ToEmailReceipt(val orderId: Long) : NavigationEvent()
+        data object ReturnHomeFromCashPayment : NavigationEvent()
```
I'd consider calling it something specific, `ReturnHomeFromCashWhenCardPaymentStarted`, to make it clear that it not just back click

samiuelson replied:
Renamed – 929c42f688625820931fca1075a87ef673e97dff

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -278,9 +278,10 @@ class WooPosTotalsViewModel @Inject constructor(
-                    is ParentToChildrenEvent.OrderSuccessfullyPaid -> showSuccessfulPaymentState(
-                        event.paymentMethod
-                    )
+                    is ParentToChildrenEvent.OrderSuccessfullyPaid -> {
```
Just checking that it's ok that cancelation will be called also in the card reader payments cases

samiuelson replied:
The flow works correctly, but I switched to canceling payment intent only in case the payment success comes from cash payment to avoid redundant logic – 60ef6e2b58d1471daea64acbe5b98b402a8524ca

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeNavigation.kt:
```diff
@@ -30,6 +30,15 @@ fun NavController.navigateToHomeScreenAfterSuccessfulCashPayment() {
     }
 }
 
+fun NavController.navigateToHomeScreenIfHomeScreenNotOpen() {
```
Just thinking, Ideally, to not even send the event if cash payment has not started 🤔  and somehow to determine this without adding any flag or something

samiuelson replied:
I think it's impossible to avoid sending redundant events unless we inject NavController (or it's observable destination state) into ViewModel, which would not be great because NavController is tightly coupled to UI layer. Let me know if you have idea how to improve it.

kidinov reviewed (commented):
Looks and works fine! I just left a few code improvements suggestions, please take a look!

samiuelson committed:
Merge branch 'trunk' into pos-cash-card-payment-collection-alignment

samiuelson committed:
Move `WooPosHomeViewModel.navigationEvent` collection up to `WooPosRootHost`

samiuelson committed:
Remove unused imports

samiuelson committed:
Rename nav event

samiuelson committed:
Cancel payment intent only if payment success comes from cash screen

samiuelson committed:
Simplify passing nav event from Totals

samiuelson committed:
Satisfy detekt's complaints

samiuelson commented:
> Looks and works fine! I just left a few code improvements suggestions, please take a look!

Thanks for the review, @kidinov! I've addressed comments and pushed improvements. Let me know what you think.

samiuelson committed:
Merge branch 'trunk' into pos-cash-card-payment-collection-alignment

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/root/navigation/WooPosRootHost.kt:
```diff
@@ -1,21 +1,43 @@
     onNavigationEvent: (WooPosNavigationEvent) -> Unit
 ) {
+    val homeViewModel = hiltViewModel<WooPosHomeViewModel>()
+    LaunchedEffect(Unit) {
```
💡 This coroutine stays active even if we navigate out of the Home screen, allowing to have navigation events handling code in single place.

kidinov replied:
Nice, good solution

kidinov reviewed (approved):
Looks and tests good!
